sway/desktop/output.c

```diff
@@ -524,18 +548,84 @@ static void damage_handle_frame(struct wl_listener *listener, void *data) {
 	pixman_region32_init(&damage);
 	if (!wlr_output_damage_attach_render(output->damage,
 			&needs_frame, &damage)) {
-		return;
+		return 0;
 	}
 
 	if (needs_frame) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+
 		output_render(output, &now, &damage);
 	}
 
 	pixman_region32_fini(&damage);
 
-frame_done:
+	return 0;
+}
+
+static void damage_handle_frame(struct wl_listener *listener, void *user_data) {
+	struct sway_output *output =
+		wl_container_of(listener, output, damage_frame);
+	if (!output->enabled || !output->wlr_output->enabled) {
+		return;
+	}
+
+	// Compute predicted milliseconds until the next refresh. It's used for
+	// delaying both output rendering and surface frame callbacks.
+	int msec_until_refresh = 0;
+
+	if (output->max_render_time != 0) {
+		struct timespec now;
+		clockid_t presentation_clock
+			= wlr_backend_get_presentation_clock(server.backend);
+		clock_gettime(presentation_clock, &now);
+
+		const long NSEC_IN_SECONDS = 1000000000;
+		struct timespec predicted_refresh = output->last_presentation;
+		predicted_refresh.tv_nsec += output->refresh_nsec % NSEC_IN_SECONDS;
+		predicted_refresh.tv_sec += output->refresh_nsec / NSEC_IN_SECONDS;
+		if (predicted_refresh.tv_nsec >= NSEC_IN_SECONDS) {
+			predicted_refresh.tv_sec += 1;
+			predicted_refresh.tv_nsec -= NSEC_IN_SECONDS;
+		}
+
+		// If the predicted refresh time is before the current time then
+		// there's no point in delaying.
+		//
+		// We only check tv_sec because if the predicted refresh time is less
+		// than a second before the current time, then msec_until_refresh will
+		// end up slightly below zero, which will effectively disable the delay
+		// without potential disasterous negative overflows that could occur if
+		// tv_sec was not checked.
+		if (predicted_refresh.tv_sec >= now.tv_sec) {
+			long nsec_until_refresh
+				= (predicted_refresh.tv_sec - now.tv_sec) * NSEC_IN_SECONDS
+					+ (predicted_refresh.tv_nsec - now.tv_nsec);
+
+			// We want msec_until_refresh to be conservative, that is, floored.
+			// If we have 7.9 msec until refresh, we better compute the delay
+			// as if we had only 7 msec, so that we don't accidentally delay
+			// more than necessary and miss a frame.
+			msec_until_refresh = nsec_until_refresh / 1000000;
+		}
+	}
+
+	int delay = msec_until_refresh - output->max_render_time;
+
+	// If the delay is less than 1 millisecond (which is the least we can wait)
+	// then just render right away.
+	if (delay < 1) {
+		output_repaint_timer_handler(output);
+	} else {
+		output->wlr_output->block_idle_frame = true;
```

Comment:
Instead of having this workaround in wlroots, I'd be more comfortable having it in Sway only, until we find a better solution. Do you think adding a `sway_output.block_frame_events` field and a `if (output->block_frame_events) { return; }` to the frame handler would be equivalent?

Reply:
I can't say if it would be equivalent from just looking at the code, since `block_idle_frame` actually blocks the call to backend's `schedule_frame`, which contains some non-trivial code in case of DRM and commits the surface right away in case of Wayland.

Reply:
I think this sequence of events happen:

* We get a frame event. We setup our repaint timer.
* A surface commits, `wlr_output_damage` schedules a frame. This sets `wlr_output.frame_pending` to true.
* Our frame timer fires, but a frame is already pending.

I think moving `block_idle_frame` into `wlr_output_damage` would be better, since it's the helper scheduling frame events behind our backs. This will help reducing the complexity added to the already-complex `wlr_output` code.

Reply:
> I think moving `block_idle_frame` into `wlr_output_damage` would be better, since it's the helper scheduling frame events behind our backs. This will help reducing the complexity added to the already-complex `wlr_output` code.

I'm worried about the following sequence of events:
1. A surface commits, a frame is scheduled using the idle timer.
2. Immediately after we get a frame event which sets up the repaint timer.
3. The idle frame renders and breaks it.

Or is the 1-2 transition like this impossible? As in, can there be a situation where a surface commit is directly followed by a frame event, without idle timers getting the chance to run in the meantime?

Reply:
>I think moving block_idle_frame into wlr_output_damage would be better, since it's the helper scheduling frame events behind our backs.

Actually, this isn't quite right. One other wlroots module calls `wlr_output_schedule_frame`: the screencopy implementation. `wlr_output.block_idle_frame` is a way for Sway to say to wlroots "don't worry, I'll send a frame shortly".

Well, let's just merge this and I'll open an issue about all of these problems. The `wlr_output` interface hasn't been designed with repaint scheduling in mind.

>I'm worried about the following sequence of events

The 1-2 transition is impossible because if a frame event is pending (e.g. because we've swapped the output buffers) then `wlr_output_schedule_frame` is a no-op.